supplierProduct/models.py

```diff
@@ -0,0 +1,67 @@
+from django.db import models
+from django.core.validators import MinLengthValidator, MinValueValidator
+from product.models import Product
+
+
+class User(models.Model):
+    user_name = models.CharField(max_length=32,
+                                 primary_key=True,
+                                 null=False,
+                                 validators=[MinLengthValidator(6)],
+                                 help_text="User Name")
+    first_name = models.CharField(max_length=32,
+                                  validators=[MinLengthValidator(1)],
+                                  help_text="First Name")
+    last_name = models.CharField(max_length=32,
+                                 validators=[MinLengthValidator(1)],
+                                 help_text="Last Name")
+    password = models.CharField(max_length=16,
+                                validators=[MinLengthValidator(6)],
+                                null=False,
+                                help_text="Password")
+    user_type = models.CharField(max_length=16,
+                                 validators=[MinLengthValidator(6)],
+                                 null=False,
+                                 choices=[('client', 'client'), ('supplier', 'supplier')],
+                                 help_text="User Type - supplier/client")
+
+    class Meta:
+        abstract = True
+
+
+class Supplier(User):
+    """Supplier class inherit from User abstract class.
+    Supplier class lets the app user to do variety of actions
+    which evolves the Product and Delivery class.
+    Attributes:
+        business_name: a string which holds the Supplier business name.
+    """
+    business_name = models.CharField(max_length=32,
+                                     validators=[MinLengthValidator(6)],
+                                     null=False,
+                                     help_text="Business Name")
+
+
+class supplierProduct(models.Model):
+    suppler_product_id = models.IntegerField(primary_key=True)
+    qr_code = models.ForeignKey(Product, on_delete=models.CASCADE)
+    user_name = models.ForeignKey('Supplier', on_delete=models.CASCADE)
+    price = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+
+    class Meta:
+        unique_together = ('suppler_product_id', 'qr_code')
+
+    def increase_Sup_product_quantiy(self):
+        self.quantity = self.quantity + 1
+        self.save()
+
+    def decrase_Sup_product_quantity(self):
+        if self.quantity == 0:
+            exit
+        else:
+            self.quantity -= 1
+            self.save()
+
+    def delete_sup_product(self):
+        supplierProduct.objects.get(suppler_product_id=self.suppler_product_id).delete()
```

Comment:
Why do we need this function? 
Also, if you have the object itself, you don't need to retrieve it again (here you could just write: `self.delete()` and achieve the same goal. But again, we don't need this).

Reply:
As i said this function can raise an error called doesnotexist(get method)
delete method only has protected error that raised when protected field is there but all we have are on cascade
do we the user deletes something that does not exist for example and he didnt get no thing he think that he deletes something but behind the scenes nothing happends,but when you get an excelption for example then its another story

Reply:
My point here was, that if the function accepts `self` as an argument (so it is not a static method), then there is no point in checking if the object exists.